server/public/model/oauth_outgoing_connection.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	if len(oa.ClientId) == 0 || utf8.RuneCountInString(oa.ClientId) > 255 {
+		return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.client_id.error", nil, "id="+oa.Id, http.StatusBadRequest)
+	}
+
+	if len(oa.ClientSecret) == 0 || utf8.RuneCountInString(oa.ClientSecret) > 255 {
+		return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.client_secret.error", nil, "id="+oa.Id, http.StatusBadRequest)
+	}
+
+	if len(oa.OAuthTokenURL) == 0 || utf8.RuneCountInString(oa.OAuthTokenURL) > 256 {
+		return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.oauth_token_url.error", nil, "id="+oa.Id, http.StatusBadRequest)
+	}
+
+	if err := oa.IsValidGrantType(); err != nil {
+		return err
+	}
+
+	if len(oa.Audiences) == 0 {
+		return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.audience.empty", nil, "id="+oa.Id, http.StatusBadRequest)
+	}
+
+	if len(oa.Audiences) > 0 {
+		for _, audience := range oa.Audiences {
+			if !IsValidHTTPURL(audience) {
+				return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.audience.error", nil, "id="+oa.Id, http.StatusBadRequest)
+			}
+		}
+	}
+
+	return nil
+}
+
+// IsValidGrantType validates the grant type and its parameters returning an error if it isn't properly configured
+func (oa *OAuthOutgoingConnection) IsValidGrantType() *AppError {
+	if !oa.GrantType.IsValid() {
+		return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.grant_type.error", nil, "id="+oa.Id, http.StatusBadRequest)
+	}
+
+	if oa.GrantType == GrantTypePassword && (oa.CredentialsUsername == nil || oa.CredentialsPassword == nil) {
+		return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.password_credentials.error", nil, "id="+oa.Id, http.StatusBadRequest)
+	}
+
+	if oa.GrantType == GrantTypePassword && (*oa.CredentialsUsername == "" || *oa.CredentialsPassword == "") {
+		return NewAppError("OAuthOutgoingConnection.IsValid", "model.oauth_outgoing_connection.is_valid.password_credentials.error", nil, "id="+oa.Id, http.StatusBadRequest)
+	}
+
+	return nil
+}
+
+// PreSave will set the Id if empty, ensuring the object has one and the create/update times.
+func (oa *OAuthOutgoingConnection) PreSave() {
+	if oa.Id == "" {
+		oa.Id = NewId()
+	}
+
+	oa.CreateAt = GetMillis()
+	oa.UpdateAt = oa.CreateAt
+}
+
+// PreUpdate will set the update time to now.
+func (oa *OAuthOutgoingConnection) PreUpdate() {
+	oa.UpdateAt = GetMillis()
+}
+
+// Etag returns the ETag for the cache.
+func (oa *OAuthOutgoingConnection) Etag() string {
+	return Etag(oa.Id, oa.UpdateAt)
+}
+
+// Sanitize removes any sensitive fields from the OAuthOutgoingConnection object.
+func (oa *OAuthOutgoingConnection) Sanitize() {
+	oa.ClientId = ""
+	oa.ClientSecret = ""
+}
```

Comment:
The `ClientId` is not sensitive, and may be used as a reference for the admin to determine something is set up properly. We should clean up the `CredentialsPassword` here right?